orttraining/orttraining/python/training/ortmodule/_data_observer.py

```diff
@@ -0,0 +1,205 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import onnx
+from onnx import helper
+from onnx import onnx_pb as onnx_proto
+import warnings
+from onnxruntime.training import ortmodule
+import torch
+
+
+class DataObserver(object):
+    """Configurable data observer for ORTModule."""
+
+    def __init__(self, log_steps=10):
+        self._enabled = ortmodule._defined_from_envvar("ORTMODULE_ENABLE_DATA_OBSERVER", 0, warn=True) == 1
+        self._embedding_graph_input_to_padding_idx_map = {}
+        self._loss_label_graph_input_to_ignore_idx_map = {}
+        self._stats = []
+
+        self._last_step = 0
+        self._current_step = 0
+        self._log_steps = log_steps
+
+        self._tensor_to_node_map = {}
+
+    def initialize(self, model, user_input_names):
+        """Initialize data observer."""
+        self._tensor_to_node_map.clear()
+        for node in model.graph.node:
+            for output_name in node.output:
+                self._tensor_to_node_map[output_name] = node
+
+        self._initialize_embedding_padding_inspector(model, user_input_names)
+        self._initialize_loss_label_padding_inspector(model, user_input_names)
+
+    def _initialize_embedding_padding_inspector(self, model, user_input_names):
+        """Register embedding input padding inspector.
+
+        This is used for collecting data/compute sparsity information for embedding layer.
+        """
+        if not self._enabled:
+            return
+
+        self._embedding_graph_input_to_padding_idx_map.clear()
+
+        for node in model.graph.node:
+            if not (
+                node.domain == "org.pytorch.aten" and node.op_type in ["ATen"] and node.input[1] in user_input_names
+            ):
+                continue
+
+            found = [attr for attr in node.attribute if attr.name == "operator"]
+            if not found or helper.get_attribute_value(found[0]).decode() != "embedding":
+                continue
+
+            tensor = self._get_initializer(model, node.input[2])
+            if tensor is None or tensor.data_type not in [onnx_proto.TensorProto.INT32, onnx_proto.TensorProto.INT64]:
+                continue
+
+            value = onnx.numpy_helper.to_array(tensor)
+            if value.ndim != 0:
+                warnings.warn(
+                    "Embedding padding_idx must be a scalar, but got a tensor of shape {}".format(value.shape)
+                )
+                continue
+
+            padding_idx = value.item()
+            # Negative padding_idx in ATen embedding means there is no padding.
+            if padding_idx < 0:
```

Comment:
I could be wrong, I think negative padding_idx does work for embedding.

Reply:
torch.nn.Embedding allow users to choose pass padding or none. If a padding is passed, then it will be checked and used, otherwise, -1 will be given to the corresponding Aten embeddign operator.

Reply:
You are right. Here we check the padding of Aten embedding operator, if it is -1, which means user did not pass padding, we will skip the data check.